### ISSUE TITLE
Command interceptors

### ISSUE THREAD

**Issue by hudo** [ironcev/Pragmatic#21]
This issue is just to start a discussion if it makes sense, so why and how to implement command interceptors. Or even query interceptors, something like it's already done with caching implementation.

For commands I would like to implement specific interceptors for audit, security, logging for example. Those interceptors can then continue command execution pipeline or stop it. 

Implementation can be done with decorators, where we inject the next command handler/interceptor, and responsibility of executing interceptor is to call or not call the next one in chain.

What do you think?

**Comment by ironcev** [comment]
What you call interceptors was called pipeline in "my" terminology.

During the implementation of query result caching, my first intention was to introduce a completely generic pipeline similar to one in ASP.NET.
Such pipeline would exist before and after both query and command execution.
It would allow e.g. to adjust a query or a command or even their results by adding handlers to the pipeline, again similar as in ASP.NET.

In that case, caching can be seen as a specific handler that is registered both before a query executes and after it executes.

What I soon realized was the clash between a super-mighty generic pipeline mechanism and a very concrete use case of caching a query result.
That (and the fact that developing a generic pipeline would take considerably more time) led me to the decision to introduce caching as a build-in feature and not just as a specific case of pipelining.

I’m saying this because I see the same question again. Yes I’m definitely for a pipeline (or interceptors, as you call it), because there will most likely be specific features in product that use Pragmatic that are not worth implementing as specific concepts in Pragmatic.

But for features like logging we should weigh if it makes more sense to implement it as built in feature.
The implementation can be as simple as e.g. logging in EF, where the client just have to attach a callback, or more sophisticated.
But in both cases, it would be supported as a built-in concept.

On the other hand, out of the gut, auditing and security (mentioned as examples on your list) sound like good candidates for a generic pipeline mechanism.

Bottom line: the feature definitely makes sense. I would just judge based on use cases and priority when to implement it. Currently I see support for asynchronous execution and several persistency layers running in parallel as the two major next features of Pragmatic. Also, there are some other request pending for longer time (https://github.com/ironcev/Pragmatic/issues).
